test/test_linalg.py

```diff
@@ -117,6 +117,239 @@ def run_test_case(a, b):
+        else:
+            drivers = ('gels', None)
+
+        def check_correctness(a, b, sol):
+            sol2 = a.pinverse() @ b
+            self.assertEqual(sol, sol2, atol=1e-5, rtol=1e-5)
+
+        def check_correctness_ref(a, b, res, ref):
+            def apply_if_not_empty(t, f):
+                if t.numel():
+                    return f(t)
+                else:
+                    return t
+
+            def select_if_not_empty(t, i):
+                selected = apply_if_not_empty(t, lambda x: x.select(0, i).numpy())
+                return selected
+
+            m = a.size(-2)
+            n = a.size(-1)
+            nrhs = b.size(-1)
+            a_3d = a.view(-1, m, n)
+            b_3d = b.view(-1, m, nrhs)
+
+            solution_3d = res.solution.view(-1, n, nrhs)
+            residuals_3d = apply_if_not_empty(res.residuals, lambda t: t.view(-1, nrhs))
+            rank_3d = apply_if_not_empty(res.rank, lambda t: t.view(-1))
+            singular_values_3d = apply_if_not_empty(res.singular_values, lambda t: t.view(-1, min(m, n)))
+            batch_size = a_3d.shape[0]
+
+            for i in range(batch_size):
+                sol, residuals, rank, singular_values = ref(
+                    a_3d.select(0, i).numpy(),
+                    b_3d.select(0, i).numpy()
+                )
+                # Singular values are None when lapack_driver='gelsy' in SciPy
+                if singular_values is None:
+                    singular_values = []
+                self.assertEqual(sol, select_if_not_empty(solution_3d, i), atol=1e-5, rtol=1e-5)
+                self.assertEqual(residuals, select_if_not_empty(residuals_3d, i), atol=1e-5, rtol=1e-5)
+                self.assertEqual(rank, select_if_not_empty(rank_3d, i), atol=1e-5, rtol=1e-5)
+                self.assertEqual(singular_values, select_if_not_empty(singular_values_3d, i), atol=1e-5, rtol=1e-5)
+
+        def check_correctness_scipy(a, b, res, driver, cond):
+            if TEST_SCIPY and driver not in (None, 'gels'):
+                import scipy.linalg
+
+                def scipy_ref(a, b):
+                    return scipy.linalg.lstsq(a, b, lapack_driver=driver, cond=cond)
+                check_correctness_ref(a, b, res, scipy_ref)
+
+        def check_correctness_numpy(a, b, res, driver, cond):
+            if driver in ('gelsd', 'gelss'):
+                import numpy.linalg
+
+                def numpy_ref(a, b):
+                    return numpy.linalg.lstsq(a, b, rcond=-1 if cond is None else cond)
+                check_correctness_ref(a, b, res, numpy_ref)
+
+        def check_ranks(a, ranks, cond=1e-7):
+            ranks2 = torch.matrix_rank(a, tol=cond)
+            ranks2 = ranks2.unsqueeze(0) if ranks2.dim() == 0 else ranks2
+            self.assertEqual(ranks, ranks2)
+
+        def check_singular_values(a, sv):
+            sv2 = a.svd()[1]
+            self.assertEqual(sv, sv2)
+
+        ms = [2 ** i for i in range(1, 5)]
```

Comment:
I think it's better to test the range starting from 0.
Suppose you have an `a` of shape _(0, m)_ and `b` of shape _(0, k)_, then the solution should be _zeros(m, k)_.

Reply:
2 ** 0  is 1. But I have tests for this special case.

Reply:
I have tests for this special case already, in a separate suite. This one compares with NumPy and Scipy, so I consider 3d representations of PyTorch tensors which I cannot review once they are 0-elements. So, this change of range from 0 is more work than it seems.